subcommands/keys/ca_revoke_device_ca.go

```diff
@@ -0,0 +1,235 @@
+package keys
+
+import (
+	x509Lib "crypto/x509"
+	"encoding/asn1"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/foundriesio/fioctl/x509"
+)
+
+const (
+	crlCmdAnnotation      = "crl-revoke"
+	crlCmdRevokeDeviceCa  = "revoke-device-ca"
+	crlCmdDisableDeviceCa = "disable-device-ca"
+)
+
+func init() {
+	revokeCmd := &cobra.Command{
+		Use:   "revoke-device-ca <PKI Directory>",
+		Short: "Revoke device CA from the list of CAs allowed to issue device client certificates",
+		Run:   doRevokeDeviceCa,
+		Args:  cobra.ExactArgs(1),
+		Long: `Revoke device CA from the list of CAs allowed to issue device client certificates.
+
+When the online or local device CA is revoked:
+- It is no longer possible to register new devices with client certificates issued by this device CA.
+- Existing devices with client certificates issued by this device CA can no longer connect to your FoundriesFactory.
+
+You may later re-add a revoked device CA using "keys ca update", if you still keep its certificate stored somewhere.
+Once you do it, devices with client certificates issued by this device CA may connect to your FoundriesFactory again.`,
+		Example: `
+# Revoke a local device CA by providing a (default) file name inside your PKI directory:
+fioctl keys ca revoke-device-ca /path/to/pki/dir --ca-file local-ca
+
+# Revoke two local device CAs given a full path to their files:
+fioctl keys ca revoke-device-ca /path/to/pki/dir --ca-file /path/to/ca1.pem --ca-file /path/to/ca2.crt
+
+# Revoke two device CAs given their serial numbers:
+fioctl keys ca revoke-device-ca /path/to/pki/dir --ca-serial <base-10-serial-1> --ca-file <base-10-serial-2>
+
+# Revoke a local device CA, when your factory root CA private key is stored on an HSM:
+fioctl keys ca revoke-device-ca /path/to/pki/dir --ca-file local-ca \
+  --hsm-module /path/to/pkcs11-module.so --hsm-pin 1234 --hsm-token-label <token-label-for-key>
+
+# Show a generated CRL that would be sent to the server to revoke a local device CA, without actually revoking it.
+fioctl keys ca revoke-device-ca /path/to/pki/dir --ca-file local-ca --dry-run --pretty`,
+		Annotations: map[string]string{crlCmdAnnotation: crlCmdRevokeDeviceCa},
+	}
+	caCmd.AddCommand(revokeCmd)
+	addRevokeCmdFlags(revokeCmd, "revoke")
+
+	disableCmd := &cobra.Command{
+		Use:   "disable-device-ca <PKI Directory>",
+		Short: "Disable device CA devices with client certificates issued by it can no longer be registered",
+		Run:   doRevokeDeviceCa,
+		Args:  cobra.ExactArgs(1),
+		Long: `Disable device CA devices with client certificates issued by it can no longer be registered.
```

Comment:
I'm a bit confused by this, maybe this would be correct? Not sure if even the suggestion is correct
```suggestion
		Long: `Disable device CA devices which client certificates issued by it can no longer be registered.
```

Reply:
Oopt! This is where I wrote a sentence which makes no sense to me.

How about this (an exact meaning):
```suggestion
Disable device CA, so that new devices with client certificates issued by it can no longer be registered.
```

And then, for the `revoke-device-ca` command, a better description would be this:
```diff
- Revoke device CA from the list of CAs allowed to issue device client certificates.
+ Revoke device CA, so that devices with client certificates issued by it can no longer connect to your factory.
```

Reply:
Looks good

Reply:
Fixed in force-push